### PR TITLE
Update deprecated api

### PR DESCRIPTION
**Pull request by jin-conbus**
- remove IInAppBillingService.aidl file and Use the Google Play Billing Library
- update deprecated API

### CODE REVIEW

File: library/src/main/java/com/anjlab/android/iab/v3/BillingProcessor.java [hunk 1]
```diff
@@ -782,11 +782,11 @@ public boolean consumePurchase(String productId)
 		try
 		{
 			TransactionDetails transaction = getPurchaseTransactionDetails(productId, cachedProducts);
-			if (transaction != null && !TextUtils.isEmpty(transaction.purchaseToken))
+			if (transaction != null && !TextUtils.isEmpty(transaction.purchaseInfo.purchaseData.purchaseToken))
```
Comment (serggl) [inline library/src/main/java/com/anjlab/android/iab/v3/BillingProcessor.java]: what is the purpose of this change? can you elaborate

Comment (jin-conbus) [reply library/src/main/java/com/anjlab/android/iab/v3/BillingProcessor.java]: The TransactionDetails.purchaseToken field has been deprecated. so used purchaseInfo.purchaseData.purchaseToken field.

Comment (serggl) [reply library/src/main/java/com/anjlab/android/iab/v3/BillingProcessor.java]: I bet behaviour will depend on the google play services version installed on a device.
I would do this in a non-breaking manner, checking for both old and new values

Comment (jin-conbus) [reply library/src/main/java/com/anjlab/android/iab/v3/BillingProcessor.java]: this function use TransactionDetails instance.
TransactionDetails class constructor.
public TransactionDetails(PurchaseInfo info)
	{
		purchaseInfo = info;
		productId = purchaseInfo.purchaseData.productId;
		orderId = purchaseInfo.purchaseData.orderId;
		purchaseToken = purchaseInfo.purchaseData.purchaseToken;
		purchaseTime = purchaseInfo.purchaseData.purchaseTime;
	}

so used purchaseInfo.purchaseData.purchaseToken field.

Comment (mustii82) [reply library/src/main/java/com/anjlab/android/iab/v3/BillingProcessor.java]: @serggl Can we Merge ?

**Comment by mustii82** [comment]
What about a new Release?

**Comment by serggl** [comment]
done ✅

**Comment by DennyWeinberg** [comment]
Anything to pay attention to? Just update to 45? I read something about "Acknowledge purchases" that needs to be done in v3?

**Comment by jin-conbus** [comment]
@DennyWeinberg This issue solves the problem in the screenshot.
image

#406 This issue maybe not application issue. Because this library is using version 3 or higher. 
**Declaration** : [Constants.GOOGLE_API_VERSION](https://github.com/anjlab/android-inapp-billing-v3/blob/6fa7398b6a9d1187d2c0eff9368f0d40a1210e80/library/src/main/java/com/anjlab/android/iab/v3/Constants.java)
**Use code** : [BillingProcessor](https://github.com/anjlab/android-inapp-billing-v3/blob/450b46d8d93f45978bb9ab435c12fb6fce9583cd/library/src/main/java/com/anjlab/android/iab/v3/BillingProcessor.java)

I manage a lot of projects, and those warnings were mostly APIs that the server calls.
Check version 'googleapis.com/androidpublisher/{version}/applications' from caller.

**Comment by DennyWeinberg** [comment]
@jin-conbus, thanks! So I think we all are not talking about the same topic, after checking #406 and #160. I'm not sure if you could leave a comment in the open issues to clarify and to make understand everybody whats going on...?
If I understand right, this library already uses v3, so we are fine for end of 2019 and we don't need to worry. The merged pull request was about another topic.

Thanks again!

**Comment by stereomatch** [comment]
@DennyWeinberg @jin-conbus @mustii82


DennyWeinberg is refering to this issue discussed on r/androiddev in his comment above:

>Thanks! Anything to pay attention to? Just update to 45? I read something about "Acknowledge purchases" that needs to be done in v3?


The question is, if anjlab is now using Google Play Billing Library version 2.0 (in x.45 ?), then will we need to change our code to acknowlege purchases explicitly.

Or is anjlab doing that somewhere already - if not, the usage guidelines should be updated, so a user can change their code that uses anjlab accordingly.


References:

https://www.reddit.com/r/androiddev/comments/bmvvey/in_app_billing_aidl_deprecated/
In App Billing AIDL deprecated
Tolriq

https://www.reddit.com/r/androiddev/comments/bsgfx6/play_store_automatic_refunds_after_72_hours/
Posted byu/Andriensis
Play Store automatic refunds after 72 hours?


https://developer.android.com/google/play/billing/billing_library_releases_notes

>Version 2.0 of the Google Play Billing library is now available. This version contains the following changes.
>Purchases must be acknowledged within three days

>Pending transactions
>With version 2.0 of the Google Play Billing library, you must support purchases where additional action is required before granting entitlement.

**Comment by stereomatch** [comment]
Also just to be clear about versions, and under what conditions the AIDL error was appearing for some developers:

- Google Play Billing library is version 2.0.x (latest)

- anjlab v3 1.0.45 now uses Google Play Billing library version 2.0.x, and does not use the AIDL way of doing things (this therefore avoids the error message some were getting about the use of AIDL)

I was not getting any error about use of AIDL when using:
anjlab-iabv3-1.0.44.jar

but the app was targeting API 27 - perhaps the AIDL error starts appearing when you target API 28 ?

Or it appears if some developers were using some specific API call in anjlab ?

**Comment by stereomatch** [comment]
@DennyWeinberg @jin-conbus @mustii82

Also just to be clear about versions, and under what conditions the AIDL error was appearing for some developers:

- Google Play Billing library is version 2.0.x (latest)

- anjlab v3 1.0.45 now uses Google Play Billing library version 2.0.x, and does not use the AIDL way of doing things (this therefore avoids the error message some were getting about the use of AIDL)

I was not getting any error about use of AIDL when using:
anjlab-iabv3-1.0.44.jar

but the app was targeting API 27 - perhaps the AIDL error starts appearing when you target API 28 ?

Or it appears if some developers were using some specific API call in anjlab ?

**Comment by jin-conbus** [comment]
@stereomatch 
What you're talking about is likely to be talking about using 'BillingClient'.
However, we are currently using IInAppBillingService.
We can use it as it is.

use 'com.android.billingclient:billing:2.0.1' only bottom reason
AIDL is now deprecated and will be removed in a future release.

But we can still use the old method.
Because include 'IInAppBillingService' and can use interface.

image

**Comment by stereomatch** [comment]
@jin-conbus

So if I understand it correctly, we have 3 different versions that maybe relevant to anjlab ?

- anjlab 1.0.44 and 1.0.45 (recently updated)

- Google Play Billing library 1.2.2 and 2.0 - are both anjlab 1.0.44 and 1.0.45 using 1.2.2 (using AIDL method, and not switched to 2.0 yet ?)

- Google Play Developer API v2 and v3 - I don't know why this is relevant to anjlab, but maybe anjlab uses some APIs from this library



The Google Play Developer API v2/v3 issue seems to be related to these threads:

https://github.com/anjlab/android-inapp-billing-v3/issues/402
Changes to the Google Play Developer API #402

https://github.com/idehub/react-native-billing/issues/160
Update to Google Play Developer API v3 (v2 end of life Dec 2019) #160



The Google Play Billing library 1.2.2 is being used by both anjlab 1.0.44 and 1.0.45 ?  That is, both use the AIDL method ?

https://developer.android.com/google/play/billing/billing_library_releases_notes
Google Play Billing Library

>Google Play Billing Library 2.0 Release (2019-05-07)
>
>Google Play Billing Library 1.2.2 Release (2019-03-07)


I ask because your first comment above seems to suggest otherwise:

>remove IInAppBillingService.aidl file and Use the Google Play Billing Library
>update deprecated API



You also said:

>What you're talking about is likely to be talking about using 'BillingClient'.
>However, we are currently using IInAppBillingService.
>We can use it as it is.
>
>use 'com.android.billingclient:billing:2.0.1' only bottom reason
>AIDL is now deprecated and will be removed in a future release.
>
>But we can still use the old method.
>Because include 'IInAppBillingService' and can use interface.


So "IInAppBillingService" is only used when you are using the AIDL method (and Google Play Billing library 1.2.2) - correct ?

Will anjlab be updated to stop using the AIDL method (and "IInAppBillingService"), and switch to Google Play Billing library 2.0 (which doesn't use AIDL) ?

**Comment by jin-conbus** [comment]
@stereomatch 

So if I understand it correctly, we have 3 different versions that maybe relevant to anjlab ? : 
- Before anjlab 1.0.45, not use the Google Play Billing library. And anjlab 1.0.45 also uses AIDL interface.

The Google Play Developer API v2/v3 issue seems to be related to these threads :
- As I mentioned above, those warnings were mostly APIs that the server calls.
Check version 'googleapis.com/androidpublisher/{version}/applications' from caller.

The Google Play Billing library 1.2.2 is being used by both anjlab 1.0.44 and 1.0.45 ? That is, both use the AIDL method ? :
- anjlab library use AIDL method.

I ask because your first comment above seems to suggest otherwise :
- I understood why you asked. Until the AIDL is removed from the Google Play Billing Library, IinappBillingService.aidl is removed to use the latest AIDL method and the AIDL method provided by the Google Play Billing Library is used. (The update deprecated API refers to the portion deprecated in anjlab as mentioned above.)

So "IInAppBillingService" is only used when you are using the AIDL method (and Google Play Billing library 1.2.2) - correct ? : 
- Yes

Will anjlab be updated to stop using the AIDL method (and "IInAppBillingService"), and switch to Google Play Billing library 2.0 (which doesn't use AIDL) ? : 
- Perhaps it will.

**Comment by stereomatch** [comment]
Thanks.

So clarifying further:

- anjlab 1.0.44, and even 1.0.45 (which fixes the error with "IinappBillingService"), STILL uses the AIDL method, but a newer variant of it, which avoids the "IinappBillingService" error.

- anjlab 1.0.44, and even 1.0.45 stil uses Google Play Billing library 1.2.x (and not 2.0 yet)

- a future version of anjlab may start to support Google Play Billing library 2.0 - at which point anjlab will stop using the AIDL method

- the "Google Play Developer API v2/v3 issue" is a separate issue, unrelated to the eventual transition to Google Play Billing library 1.2.x (AIDL) to 2.0

**Comment by xDragonZ** [comment]
So this PR is still using aidl and purchase no need to be acknowledged yet?

**Comment by stereomatch** [comment]
>So this PR is still using aidl and purchase no need to be acknowledged yet?

@xDragonZ that is my impression.

**Comment by virusman** [comment]
@jin-conbus I'm confused. This PR is _already_ using Google Play Billing library 2.0. How is it using AIDL instead if Google Play Billing 2.0 doesn't use AIDL?
